internal/consul/certmanager.go

```diff
@@ -246,6 +250,25 @@ func (c *CertManager) Manage(ctx context.Context) error {
 		if err := w.RunWithClientAndHclog(c.consul, c.logger); err != nil {
 			c.logger.Error("consul watch.Plan returned unexpectedly", "error", err)
 		}
+		c.logger.Trace("consul watch.Plan stopped")
+	}
+
+	// Consul 1.11 has a bug where blocking queries on the leaf certificate endpoint
+	// cause all subsequent non-blocking queries to unexpectedly block. The problem
+	// is that this means that, on restart, the query for a leaf certificate with
+	// the given service id will never return until the previous leaf certificate
+	// expires/is rotated. Adding a wait here causes the API to return once the timeout has
+	// been hit -- allowing us to short-circuit the buggy blocking. The subsequent
+	// goroutines can then be leveraged to pick up any certificate rotations.
+	if !c.skipExtraFetch {
+		leafCert, _, err := c.consul.Agent().ConnectCALeaf(c.service, &api.QueryOptions{
+			WaitTime: 1 * time.Second,
+		})
```

Comment:
Wondering if churning / repeated calls here would be possible.

I don't have a lot of context on the project here btw, just a thought.

Reply:
Just to make sure I'm understanding -- is it a question about whether or not we'd repeatedly call this code or a suggestion to wrap this in a retry loop?

If the former, this code path should get called exactly once at the beginning of the process leveraging it (in this case our controller and gateway instances which use the CA endpoints to establish a trusted SDS connection), the goroutines that get spawned after this leverage `api/watch` to make repeated blocking query-based updates when the connect certs rotate.

Since the bug is really only a problem when we first boot up due to hanging when `api/watch` first kicks off its non-blocking request (which become blocking requests due to the bug), we essentially just jam this here to do a one-off retrieval before the blockng kicks in.

Reply:
My understanding from the discussion was "if the RPC call takes longer than the timeout (even if it doesn’t block) it could return an emtpy result" so it would be important to retry here if needed?

Reply:
So, in this particular case it I believe it should still be fine -- if the result is empty due to a timeout, the background watch should pick up the certificate since it has no timeout on it (hence why the non-blocking call became blocking previously). I know there are some oddities around this working the way it does, but it is a workaround v. ideally just using the watch code.